### ISSUE TITLE
[Feature]: Replay-based self-improvement for Hermes exploration policies (Dream-RSI)

### ISSUE THREAD

**Issue by MaxSchoon** [NousResearch/hermes-agent#117030]
## Problem or Use Case

Hermes already has meaningful self-improvement at the memory and skill layers, but it does not appear to have a mechanism that improves the **general exploration/orchestration policy itself** from accumulated execution history.

The recent [Dream-RSI paper](https://arxiv.org/abs/2609.14858) makes that layer explicit. Its core idea is:

```text
online exploration
    ↓
historical discovery trees
    ↓
replay simulator
    ↓
off-policy evaluation of alternative exploration policies
    ↓
improved policy
    ↓
redeploy online and collect better/larger discovery trees
    ↺
```

The underlying coding agent stays fixed. What changes is the lightweight orchestration policy that decides how search is conducted.

In Hermes terms, an exploration policy could govern bounded decisions such as:

- whether to continue, stop, retry, or backtrack;
- whether to branch/delegate and how many branches to create;
- how to allocate a finite tool/token/time budget across branches;
- when to request an independent verifier or stronger model;
- when to prune an unproductive branch and concentrate compute elsewhere.

This should **not** govern approvals, credentials, filesystem/network authority, tool permissions, or other safety boundaries.

Hermes already has much of the substrate needed for an experiment like this:

| Existing work | What it already provides | Why this proposal is different |
| --- | --- | --- |
| #71266 | Validation-gated self-improvement of `SKILL.md` from past sessions | Optimizes procedural knowledge, not the general search/orchestration policy |
| #113008 / #113020 | Shadow mode, replay, telemetry, and evaluation for bounded semantic decisions | Evaluates individual decision families, not long-horizon exploration strategy |
| #82253 | Repeatable black-box evaluation harness for skill activation | Measures one behavioral surface rather than optimizing exploration |
| #93305 | Replayable environments, rewards/verifiers, trajectory capture | Useful environment substrate, but no policy-improvement loop |
| #83193 / #72082 | Lifecycle and scope problems around background self-improvement | Important guardrails: policy improvement must have explicit timing, provenance, and write authority |

Hermes also already saves trajectories through `agent/trajectory.py` / `batch_runner.py`. The missing piece is a generic way to turn those histories into **replayable exploration trees** and use them to evaluate a candidate orchestration policy before spending more live inference.

I searched the issue and PR tracker for Dream-RSI, recursive self-improvement, trajectory replay, exploration policy, orchestration policy replay, and related terms. I found the adjacent work above, but not an issue proposing this general policy-level loop.

## Proposed Solution

Add an **opt-in, versioned exploration-policy layer with offline replay evaluation**. This should adopt the Dream-RSI methodology natively using Hermes primitives, not add a dependency on a specific third-party implementation.

### 1. Record replayable exploration trees

Extend trajectory/session instrumentation with a compact exploration event schema. A decision point should record enough information to reconstruct the realized search tree, for example:

```text
decision_id
parent_id
policy_version
state/progress summary or digest
available bounded actions
chosen action
child/delegation ids
verifier/outcome
token/tool/time cost
terminal status
```

Raw secrets and unrelated session context should not be copied into this log by default. Full artifacts can remain referenced through existing session/trajectory provenance.

### 2. Introduce a bounded `ExplorationPolicy` seam

Illustrative interface only:

```python
class ExplorationPolicy(ABC):
    def decide(self, state: ExplorationState) -> ExplorationDecision:
        ...
```

The decision schema should be deliberately narrow, e.g. continuation, branch count, retry/backtrack, verification, delegation, and budget allocation.

A policy must only be able to **narrow or schedule already-authorized capabilities**. It must never broaden toolsets, bypass approvals, change credential access, or override operator policy.

Prefer a versioned, inspectable policy artifact over unconstrained self-editing of Hermes core code. The artifact could be declarative or a tightly scoped module, but it should be diffable, attributable, rollbackable, and testable.

### 3. Add a replay simulator over realized discovery trees

Given historical exploration trees, evaluate a candidate policy without re-running every expensive online branch.

The simulator should be conservative about counterfactuals:

- if the candidate chooses a branch/action that exists in the recorded tree, replay its known downstream outcome;
- if it chooses an action that was never explored, mark that decision **unscorable / out of support** rather than inventing an outcome;
- report replay coverage alongside the score so a policy cannot look good merely because difficult counterfactuals were skipped.

Useful objective components could include:

- verifier/task success;
- tokens and tool calls;
- wall-clock latency;
- number of branches/delegations;
- retry/error count;
- optional task-specific reward where a deterministic evaluator exists.

This is the key distinction from simply asking an LLM to reflect on old transcripts. The candidate gets cheap, measurable feedback before another live deployment.

### 4. Generate candidates, then gate them on held-out history

A strong auxiliary model may propose a bounded policy edit, but the edit should not be promoted merely because its explanation sounds plausible.

Suggested flow:

```text
current policy
    ↓
candidate proposal
    ↓
replay on training discovery trees
    ↓
replay on held-out discovery trees
    ↓
reject / stage / shadow
    ↓
limited online canary
    ↓
promote or roll back
```

Promotion should require a configured measurable improvement while respecting hard regression limits. A rejected candidate should be retained with its evaluation result so the optimizer does not repeatedly rediscover the same dead end.

### 5. Make deployment staged and reversible

The initial implementation should change no default Hermes behavior.

A candidate policy should have:

- a content hash/version;
- parent policy version;
- replay metrics and coverage;
- provenance for the model/run that proposed it;
- shadow/canary status;
- one-command rollback;
- optional human approval before promotion.

New online runs then expand the discovery-tree pool, creating the recursive loop described by Dream-RSI.

### 6. Reuse existing Hermes evaluation surfaces

Where possible, this should compose with rather than duplicate:

- trajectory/session persistence for history;
- #113008/#113020 style shadow/replay telemetry for bounded decisions;
- #93305 style deterministic environment verifiers where available;
- #71266 style staged adoption and rejected-candidate memory;
- existing delegation/subagent machinery for controlled online exploration.

The optimizer itself could remain optional or plugin-provided. The core value would be the generic **trace → policy → replay → staged promotion** contract.

### Acceptance criteria

- [ ] Default Hermes behavior is unchanged unless exploration-policy learning is explicitly enabled.
- [ ] A Hermes task can emit a versioned exploration tree with parent/child decisions and outcome/cost metadata.
- [ ] A candidate policy can be evaluated offline against recorded trees without invoking the underlying task model.
- [ ] Unsupported counterfactual actions are reported as out-of-support rather than assigned invented rewards.
- [ ] Replay reports both score and coverage.
- [ ] Candidate and baseline are evaluated with the same underlying model/environment assumptions.
- [ ] A policy artifact is inspectable, versioned, diffable, and rollbackable.
- [ ] Policy decisions cannot broaden tool permissions or bypass approvals/safety boundaries.
- [ ] A candidate can run in shadow/canary mode before promotion.
- [ ] Promotion/rejection records objective metrics and provenance.
- [ ] At least one deterministic or safely replayable benchmark demonstrates the full loop: history → candidate → replay gate → online canary → promote/reject.
- [ ] No model weight training or fine-tuning is required.

### Extend skill self-improvement only (#71266)

This is valuable and complementary. A better `SKILL.md` can improve how one procedure is executed, but it does not learn general rules such as when to branch, how much parallel exploration to use, when to stop a branch, or when verification is worth the cost.

### Use online A/B testing only

Online canaries are still necessary because replay is limited to the realized search space, but using live runs as the only feedback signal makes policy search slow and expensive. Dream-RSI's useful contribution is extracting cheap feedback from accumulated histories first.

### Fine-tune or RL-train the underlying model

That optimizes a different layer and requires a training pipeline. The interesting Hermes opportunity is to keep the model fixed and improve how inference compute is orchestrated around it.

### Let the agent freely rewrite Hermes core

This is closer to literal recursive self-modification, but it makes validation, rollback, safety boundaries, and attribution much harder. A bounded exploration-policy seam captures most of the research idea while remaining reviewable.

### Implement the whole feature as a standalone plugin

A plugin is a good place for the optimizer or experimental policy implementations. However, generic access to replayable exploration events, bounded orchestration decisions, and shadow/promotion semantics likely needs a stable host surface. If maintainers prefer a plugin-first proof of concept, the core request could be reduced to exposing those generic seams.

## Feature Type

Performance / reliability

## Scope

Large (new module or significant refactor)

This can still be delivered incrementally: instrumentation first, replay second, policy interface third, and automated candidate generation/promotion last.

## Contribution

- [ ] I'd like to implement this myself and submit a PR

## Debug Report (optional)

N/A. This is an architectural feature proposal, not an environment-specific failure.
